src/cursor/core_cursor.ts

```diff
@@ -3,139 +3,189 @@ import { ReadPreference } from '../read_preference';
+}
+
+export interface DisconnectHandler {
+  add?(): void;
+  execute?(): void;
+  flush?(): void;
+  addObjectAndMethod(...args: unknown[]): void;
 }
 
-export type CoreCursorOptions = BSONSerializeOptions;
+export interface StreamOptions {
+  /** A transformation method applied to each document emitted by the stream */
+  transform?(doc: Document): Document;
+}
+
+export interface InternalCursorState extends BSONSerializeOptions {
+  postBatchResumeToken?: ResumeToken;
+  batchSize: number;
+  cmd: Document;
+  currentLimit: number;
+  cursorId: Long | null;
+  lastCursorId?: Long | null;
+  cursorIndex: number;
+  dead: boolean;
+  killed: boolean;
+  init: boolean;
+  notified: boolean;
+  documents: Document[];
+  limit: number;
+  operationTime?: OperationTime;
+  reconnect?: boolean;
+  session?: ClientSession;
+  skip: number;
+  streamOptions?: StreamOptions;
+  transforms?: TransformFunctions;
+  raw?: boolean;
+}
 
 // Possible states for a cursor
-const CursorState = {
-  INIT: 0,
-  OPEN: 1,
-  CLOSED: 2,
-  GET_MORE: 3
-};
-
-interface CoreCursor {
-  close(options?: any, callback?: Callback): Promise<void> | void;
+export enum CursorState {
+  INIT = 0,
+  OPEN = 1,
+  CLOSED = 2,
+  GET_MORE = 3
 }
 
-/**
- * This is a cursor results callback
- *
- * @callback resultCallback
- * @param {error} error An error object. Set to null if no error present
- * @param {object} document
- */
+export interface CoreCursorOptions extends BSONSerializeOptions {
+  noCursorTimeout?: boolean;
+  tailable?: boolean;
+  raw?: boolean;
+  hint?: Hint;
+  disconnectHandler?: DisconnectHandler;
+  limit?: number;
+  skip?: number;
+  session?: ClientSession;
+  readPreference?: ReadPreference;
+  /** The number of documents to return per batch. See {@link https://docs.mongodb.com/manual/reference/command/find/| find command documentation} and {@link https://docs.mongodb.com/manual/reference/command/aggregate|aggregation documentation}. */
+  batchSize?: number;
+  /** Initial documents list for cursor */
+  documents?: Document[];
+  /** Transform function */
+  transforms?: TransformFunctions;
+}
 
 /**
  * The **CoreCursor** class is an internal class that embodies a cursor on MongoDB
  * allowing for iteration over the results returned from the underlying query.
  *
  * **CURSORS Cannot directly be instantiated**
- *
- * @property {number} cursorBatchSize The current cursorBatchSize for the cursor
- * @property {number} cursorLimit The current cursorLimit for the cursor
- * @property {number} cursorSkip The current cursorSkip for the cursor
  */
-class CoreCursor extends Readable {
-  operation: any;
-  pool: any;
-  server: any;
-  disconnectHandler: any;
-  ns: any;
-  namespace: any;
-  cmd: any;
-  options: any;
-  topology: any;
+export class CoreCursor<
+  T extends CoreCursorOptions = CoreCursorOptions,
+  O extends OperationBase = OperationBase
+> extends Readable {
+  operation?: O;
+  pool!: ConnectionPool;
+  server!: Server;
```

Comment:
It's assigned in _initializeCursor, which does say: '// NOTE: this goes away once cursors use `executeOperation`'

Reply:
Oh I see, no this is necessary. I think it should be a `server?` which could actually second as a check for whether the cursor has been initialized. We pin to a server once the cursor begins iteration, so that's why its saved off here